[chashaobao] iP

chashaobao123 opened the pull request:
# LittleMissHelpful

> “Your mind is for having ideas, not holding them.” – David Allen

**Ah Bang Mang** frees your mind from remembering the things you need to do. It's:

- **Text-based**
- **Easy to learn**
- **SUPER FAST** to use
- **Singlish Ui**

All you need to do is:

1. [Download it from here](#).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉.

And the best part? It’s **FREE**!

## Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use LittleMissHelpful to practice Java too! Here's the main method:

```java
public static void main(String[] args) {
        new LittleMissHelpful("data/LittleMissHelpful.txt").run();
    }

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
ip_wk2_level-0

Commit:
ip wk2 level-1

Commit:
ip wk2 level-2

Commit:
wk2 ip level-3

Commit:
wk2 level-4

Commit:
wk2 ip level-5

Commit:
wk2 ip level-6

Commit:
add automated-text-ui-testing

Commit:
add Level-7 changes

AppleJem commented on src/main/java/LittleMissHelpful.java:
```diff
@@ -0,0 +1,241 @@
+        loadTasks(list);
+
+        int counter = list.size();;
+        while (true) { // Continuous input
```
Would it be better to refactor this part into the the Parse class as suggested in the project brief week 3?

kimberlytmq replied:
agreed! Separating all the parsing commands into a separate parse class could make the code look neater

chashaobao123 replied:
Yes, thank you! I have abstracted out the classes now

AppleJem commented on src/main/java/LittleMissHelpful.java:
```diff
@@ -0,0 +1,241 @@
+                        System.out.println("Ok, I see you laze. You have marked this task as not done yet: " + curTask.toString());
+                        System.out.println(lineBreak);
+                    } catch (NumberFormatException e) {
+                        System.out.println(lineBreak + "\nHai-yo, task number must be an integer lah.\n" + lineBreak);
```
This check looks the same as the check for the "mark" command, perhaps can refactor to reduce repetition?

AppleJem commented on src/main/java/LittleMissHelpful.java:
```diff
@@ -0,0 +1,241 @@
+                    Task newTask = null;
+
+                    try {
+                        if (command.equalsIgnoreCase("todo")) {
```
The switch and case statement may come in handy here. (frankly I also never use)

AppleJem commented on src/main/java/LittleMissHelpful.java:
```diff
@@ -0,0 +1,241 @@
+                } else {
+                    // Invalid command --> throw exception
+                    System.out.println(
+                        lineBreak + "\nAlamak! Format salah already... Please provide a valid command and description.\n" + lineBreak);
```
Perhaps can provide some descriptions of what are valid commands

AppleJem commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,18 @@
+public class Deadline extends Task {
+    protected String by;
```
Would it make sense to save the string as Java Datetime instead?

Commit:
update reading and writing of saved tasks

Commit:
add date and time for deadline and event

kimberlytmq commented on src/main/java/LittleMissHelpful.java:
```diff
@@ -0,0 +1,245 @@
+public class LittleMissHelpful {
+    private static final String FILE_PATH = "./data/LittleMissHelpfullist.txt";
+    public static void main(String[] args) {
+        /**
```
Maybe you can add this comment in the header instead? (i.e. before the public class in Line 10) It may help others to understand your code purpose better!

kimberlytmq commented on src/main/java/LittleMissHelpful.java:
```diff
@@ -0,0 +1,245 @@
+
+                if (inputList.length < 2) {
+                    // Check if the user command is "bye" -> exit
+                    if (command.equalsIgnoreCase("bye")) {
```
Maybe you could separate the UI related code to a UI class?

kimberlytmq commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,112 @@
+public class Task {
```
Could you consider adding header comments to the classes? Saw this problem in some other classes too.

kimberlytmq commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,112 @@
+    protected String description;
+    protected boolean isDone;
+
+    public Task (String description) {
```
An idea is that you could make this into an abstract class and have the different methods for the different classes in their respective class files! It's just an alternative way to do it that you can consider.

Commit:
abstract more classes

Commit:
abstract out classes

frymash commented on src/main/java/LittleMissHelpful/TaskList.java:
```diff
@@ -0,0 +1,59 @@
+package main.LittleMissHelpful;
+import java.util.ArrayList;
```
Small suggestion involving code consistency but I noticed some files have a line of space between the `package` statement and the `import` statement while other files don't. The course's coding standard doesn't really prescribe a set way to go about this, but perhaps you could consider either using one or the other.

chashaobao123 replied:
Thank you for your feedback! I'll be sure to edit this

frymash commented on src/main/java/LittleMissHelpful/Ui.java:
```diff
@@ -0,0 +1,87 @@
+import main.LittleMissHelpful.Task.Task;
+
+public class Ui {
+    private static final String LINE_BREAK = "---------------------------------";
```
Good practice to make the divider a constant so that you don't have to repeat yourself whenever you use it.

frymash commented on src/main/java/LittleMissHelpful/Task/Task.java:
```diff
@@ -0,0 +1,121 @@
+         */
+    
+        String[] parts = taskString.split(" \\| ");
+        if (parts.length < 3) {
+            throw new InvalidTaskFormatException("Invalid task format: " + taskString);
+        }
+    
+        String taskType = parts[0];
+        boolean isDone = parts[1].equals("1");
+        String description = parts[2];
+        
+        // Check task type and create corresponding Task
+        if (taskType.equals("T")) {
+            return new Todo(description, isDone);
+
+        } else if (taskType.equals("D")) {
+            if (parts.length != 4) {
+                throw new InvalidTaskFormatException("Invalid Deadline format: " + taskString);
+            }
+            LocalDateTime by = LocalDateTime.parse(parts[3]);
+            return new Deadline(description, isDone, by);
+
+        } else if (taskType.equals("E")) {
+            if (parts.length != 5) {
+                throw new InvalidTaskFormatException("Invalid Event format: " + taskString);
+            }
+
+            LocalDateTime from = LocalDateTime.parse(parts[3]);
+            LocalDateTime to = LocalDateTime.parse(parts[4]);
+            return new Event(description, isDone, from, to);
```
Could possibly refactor the if-else statement into a switch-case statement (not 100% necessary but a suggestion).

frymash commented on src/main/java/LittleMissHelpful/Task/Task.java:
```diff
@@ -0,0 +1,121 @@
+        LocalDateTime from = null;
+        LocalDateTime to = null;
+
+        if (this instanceof Todo) {
+            taskType = "T";
+            description = this.description;
+            return taskType + spacer + isDone + spacer + description;
+
+        } else if (this instanceof Deadline d) {
+            taskType = "D";
+            description = d.description;
+            by = d.by;
+            return taskType + spacer + isDone + spacer + description + spacer + by;
+
+        } else if (this instanceof Event e) {
+            taskType = "E";
+            description = e.description;
+            from = e.from;
+            to = e.to;
+            return taskType + spacer + isDone + spacer + description + spacer + from + spacer + to;
+        } else {
+            throw new InvalidTaskFormatException("Cannot write task: " + this.toString());
```
Was wondering if the switch-case syntax would be more idiomatic here too

frymash reviewed (commented):
In general, I think your code is really well-organised and well-abstracted. This is the first time I've seen someone create a Java file for each command. I also like the vibes your bot's responses give. Good job! LGTM.

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
correct packages

Commit:
correctly add packages

Commit:
add JUnit: TaskListTest and UiTest

Commit:
add headers

Commit:
add search function

Commit:
resolve merge conflicts

Commit:
merge branches

JoanneHing commented on src/main/java/LittleMissHelpful/Command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,53 @@
+import LittleMissHelpful.Ui;
+import LittleMissHelpful.Exception.InvalidCommandException;
+import LittleMissHelpful.Task.Deadline;
+
```
The header comment for the class might be missed out, will be good if it could be added in

JoanneHing commented on src/main/java/LittleMissHelpful/Command/AddEventCommand.java:
The header comment for the class and the public methods is also missed out here, would be nice to add in!

JoanneHing commented on src/main/java/LittleMissHelpful/Command/AddEventCommand.java:
```diff
@@ -0,0 +1,70 @@
+        if (dateTimeString.equalsIgnoreCase("today")) {
+            // Use current date and set time to 00:00 (start of the day)
+            return LocalDateTime.now().withHour(00).withMinute(00);
+        }
```
line breaks could be added here to seperate the section and make the code more readabel : )

JoanneHing commented on src/main/java/LittleMissHelpful/Command/Command.java:
missing header comment for the class too

JoanneHing commented on src/main/java/LittleMissHelpful/Command/DeleteTaskCommand.java:
same as above

JoanneHing commented on src/main/java/LittleMissHelpful/Parser.java:
```diff
@@ -0,0 +1,73 @@
+        String item = inputList[1];
+        switch (command) {
+            case "mark":
+                return new MarkTaskCommand(Integer.parseInt(item));
+
+            case "unmark":
+                return new UnmarkTaskCommand(Integer.parseInt(item));
+
+            case "delete":
+                return new DeleteTaskCommand(Integer.parseInt(item));
+
+            case "find":
+                return new FindCommand(item);
+
+            case "todo":
+                return new AddTodoCommand(item);
+
+            case "deadline":
+                String[] deadlineParts = item.split("/by", 2);
+                if (deadlineParts.length < 2) {
+                    throw new InvalidCommandException("Invalid deadline format. Use 'deadline description /by date'.");
+                }
+                String deadlineDesc = deadlineParts[0].trim();
+                String deadlineBy = deadlineParts[1].trim();
+                return new AddDeadlineCommand(deadlineDesc, deadlineBy);
+
+            case "event":
+                String[] eventParts = item.split("/from | /to", 3);
+                if (eventParts.length < 3) {
+                    throw new InvalidCommandException("Invalid event format. Use 'event description /from start /to end'.");
+                }
+                String eventDesc = eventParts[0].trim();
+                String eventFrom = eventParts[1].trim();
+                String eventTo = eventParts[2].trim();
+                return new AddEventCommand(eventDesc, eventFrom, eventTo);
+
+            default:
+                throw new InvalidCommandException("Invalid command. Please provide a valid command.");
+        }
+    }
```
the `case` should not be indented, would be nice if it is fixed!

JoanneHing commented on src/main/java/LittleMissHelpful/Command/AddEventCommand.java:
```diff
@@ -0,0 +1,70 @@
+import LittleMissHelpful.Ui;
+import LittleMissHelpful.Exception.InvalidCommandException;
+import LittleMissHelpful.Task.Event;
+
```
The commands naming is named well! Easy to understand what the class is handling!

JoanneHing commented on src/main/java/LittleMissHelpful/Exception/InvalidCommandException.java:
The exceptions class name is named well to let developer understand what exception it is handling too! Well done!

JoanneHing commented on src/main/java/LittleMissHelpful/Parser.java:
```diff
@@ -0,0 +1,73 @@
+            }
+        }
+
+        String item = inputList[1];
```
Maybe item could be named to `taskIndex` for better understanding?

JoanneHing commented on src/main/java/LittleMissHelpful/Task/Event.java:
```diff
@@ -0,0 +1,72 @@
+        this.to = to;
+    }
+
+    // public Event(String description, LocalDate from, LocalDate to) {
+    //     /**
+    //      * Instatiate event without needing to input time
+    //      * Defaults from time to start of the day and end time to end of the day
+    //      */
+    //     super(description);
+    //     this.from = from.atStartOfDay(); // Set default time to start of day
+    //     this.to = to.atStartOfDay().plusDays(1).minusSeconds(1); // Set end of day time
+    // }
+
+    // public Event(String description, boolean isDone, LocalDate from, LocalDate to) {
+    //     /**
+    //      * Instatiate event without needing to input time && isDone
+    //      * Defaults from time to start of the day and end time to end of the day
+    //      */
+    //     super(description, isDone);
+    //     this.from = from.atStartOfDay(); // Set default time to start of day
+    //     this.to = to.atStartOfDay().plusDays(1).minusSeconds(1); // Set end of day time
+    // }
```
can consider removing unused code to improve code quality

JoanneHing reviewed (commented):
Overall LGTM! The naming for class , method and variables is well done and convey their purposes clear! 

Would be nice if header comments could be added to class and the public method as part of JavaDoc! 

Well done!

Commit:
add gui

Commit:
add gui

Commit:
Add assertions

Assertions were added to LittleMissHelpful file to enhance clarity and debugging

Let's,
* add assertions to LittleMissHelpful
* enable assertions in the configurations

Commit:
Update packages to follow convention

Packages were not in full lowercase

Let's
* rename all packages to follow convention.

Commit:
Update packages

chashaobao123 committed:
Merge pull request #2 from chashaobao123/branch-A-Assertions

Add assertions

Commit:
Resolve import errors

Commit:
Merge with master

chashaobao123 committed:
Merge pull request #3 from chashaobao123/branch-A-CodeQuality

Branch a code quality

Commit:
Add assertions

Commit:
Merge branch 'master' of github.com:chashaobao123/ip

Commit:
Merge Level-8 to master

chashaobao123 committed:
Merge pull request #4 from chashaobao123/branch-Level-8

add date and time for deadline and event

Commit:
Resolved merge conflicts from pull-request

Commit:
Add proper headings

Commit:
Edit error messages to fit personality

Commit:
Add JUnit tests

Commit:
Update README

Commit:
Update README

Commit:
Add Ui.png to docs folder

Commit:
Update README
